### ISSUE TITLE
Tracking issue for future-compatibility warning `recursion_depth_exceeding_limit`

### ISSUE THREAD

**Issue by adwinwhite** [rust-lang/rust#159228]
This is the **summary issue** for the `recursion_depth_exceeding_limit` future-compatibility warning. The goal of this page is to describe why this change was made and how you can fix code that is affected by it. It also provides a place to ask questions
or register a complaint if you feel the change should not be made. For more information on the policy around future-compatibility warnings, see our [breaking change policy guidelines](https://github.com/rust-lang/rfcs/blob/master/text/1122-language-semver.md).

### What is this warning for

The `recursion_depth_exceeding_limit` lint detects cases where the compiler previously did not correctly track the required recursion depth. Properly tracking the required depth then causes overflow errors as it exceeds the value set for the `#[recursion_limit = "N"]`. We are temporarily weakening these errors to a future compatibility warning.

#### Example

```rust
#![recursion_limit = "8"]

// The field order matters 😂
struct Foo<T> {
    t: T,
    opt_t: Option<T>,
}

fn require_sync<T: Sync>() {}

fn main() {
    // The warning: "overflow evaluating the requirement `Foo<Foo<Foo<Foo<Foo<Foo<()>>>>>>: Sync`"
    require_sync::<Foo<Foo<Foo<Foo<Foo<Foo<()>>>>>>>();
}
```

The above code compiles with the old solver, while it requires `#[recursion_limit = "12"]` for the next solver. To prove `Foo^N: Sync`, we have to prove constituent nested goals `Foo^N-1: Sync` and `Option<Foo^N-1>: Sync`. If we require depth `d` to prove the former, the next solver requires the depth `d + 1` for the later, because we goes down one depth from `Option<Foo^N-1>: Sync` to its constituent goal `Foo^N-1: Sync`, and we hit the cache with required depth d for this nested goal.
So, the total required depth for the root goal is proportional to `N` with slope `2`. But for the old solver, we check the recursion limit when we actually consider the obligation and don't record/lookup the required depth for its deeply nested goals.
Therefore, the total required depth for the root goal is proportional to `N` with slope `1` and the next-solver requires twice more recursion limits as `N` grows.

### How to best fix this

It's strongly recommended to add manual impls for auto traits like below, if possible.
```rust
// First identify the relevant auto trait, usually `Send/Sync`. It's possible that the auto trait
// is in an indirect where clause of impls rather than in the displayed clause.
// Then find the type required to implement the auto trait. The type is expected to be 
// deeply nested, requiring recursive visits into constituent fields.
//     - If the type is concrete, use `unsafe impl YourAutoTrait for YourType {}`
//     - If the type is generic, use `unsafe impl<T: YourAutoTrait> YourAutoTrait for YourType<T> {}`
// To ensure the fields of your type actually implement the auto trait, you can validate them like below.
//
// Adapted from the `non_structural_derive` crate. You may want to use it directly.
unsafe impl<T: Sync> Sync for Foo<T> {}
fn _check_bound<T: Sync>(_: &T) {}
fn _validate_fields<T: Sync>(x: &Foo<T>) {
    let Foo { t, opt_t } = x;
    _check_bound::<T>(&t);
    _check_bound::<Option<T>>(&opt_t);
}
```
This helps your downstream users avoid having to increase the recursion_limit when they use your deeply nested types. This also improves the performance of the type system when compared to the builtin auto-trait impl which may matter for very large types.

If your overflow isn't caused by auto traits, you may have to increase the `recursion_limit` by adding the following to your crate root.
```rust
#![recursion_limit = "a big number like 256"]
```
It is generally better to change the code to not require a large recursion depth as this requires all downstream users to also increase their `recursion_limit`.

### Impact

This affects `wgpu = "25.0.2"` and a lot of other crates. This was investigated and explained by @ShoyuVanilla. [zulip discussion](https://rust-lang.zulipchat.com/#narrow/channel/364551-t-types.2Ftrait-system-refactor/topic/The.20old.20solver.20don.27t.20record.20required.20depth.20on.20cache/with/604756008)

**Comment by Lorak-mmk** [comment]
We recently started to encounter this in Scylla Rust Driver ( https://github.com/scylladb/scylla-rust-driver )
This is frustrating for a few reasons:
 - It breaks in CI, but not for me locally, despite the same Rust version and env variables being used. This makes it harder to workaround.
 - All the advice around the internet suggests `#![recursion_limit = "256"]`. This is fine if the problem appears in one place, but we would have to add it to each example, and even worse: each doctest. This is not really feasible. A flag for rustc would help, because we could just add it in CI, but its unstable :(
 - I don't see how to investigate where all this depth comes from. If I knew, I could perhaps try to rewrite the affected code to decrease the depth.

What to do? How do we workaround this?

**Comment by adwinwhite** [comment]
@Lorak-mmk Do you mind filing an issue or starting a thread on [#t-types/trait-system-refactor](https://rust-lang.zulipchat.com/#narrow/channel/364551-t-types.2Ftrait-system-refactor) to dump the info you have, like environments and logs?

This is probably caused by certain types having very deep fields to recurse into. You might want to have a look at the updated [how to best fix this](https://github.com/rust-lang/rust/issues/159228#issue-4874441278) on how to add an impl to short-circuit this.

At the moment, we're working on displaying a chain of clauses in the warning which should help identify the cause.

**Comment by Lorak-mmk** [comment]
The difference between CI and my machine turned out to be due to different Cargo.lock. CI generated a new one, and I guess some dependency added a layer in a newer version.

For now, some LLM managed to find 2 places where we can use a boxed dyn Future, and cut the max depth in half thanks to that. This is not on our hot path, so it should be fine for now.

And the possible solutions (box dyn future, implementing auto traits manually) are reasonable. The only thing missing is easily seeing where all the recursion depth comes from. You said you are already working on that, so I don't really have anything new to add.

**Comment by tliron** [comment]
This is a well-meaning change, but it's causing some cascading issues for us.

Specifically they are around detecting that *futures* are `Send`, which is required by multi-threaded tokio. The problem is that there is no explicit type that we can `unsafe impl Send` because the trail of futures is all created by `async` sugar.

(Moreover, this requirement is so hard on the trait solver that we had to resort to using [AlwaysSend](https://docs.rs/always_send/latest/always_send/struct.AlwaysSend.html) in some cases.)

```rustc 1.100.0-nightly (fb6531d55 2026-08-23)```

Here's an example of one warning we get. Note that it's a bit curious, as it mixes "async fn body" and "coroutine witness" (what is that?). This warning, of course, doesn't appear in rustc stable.

```
warning: overflow evaluating the requirement `{async fn body of floria::entities::imperative::imperative::<impl floria::plugins::PluginContext<Arc<Box<dyn floria::AsyncStore + Send + Sync>>>>::set_property_checked()}: always_send::FutureExt`
  |
  = note: which requires {async fn body of floria::entities::imperative::imperative::<impl floria::plugins::PluginContext<Arc<Box<dyn floria::AsyncStore + Send + Sync>>>>::set_property_checked()}: Send
  = note: which requires {coroutine witness@floria::entities::imperative::imperative::<impl floria::plugins::PluginContext<Arc<Box<dyn floria::AsyncStore + Send + Sync>>>>::set_property_checked::{closure#0}}: Send
  = note: which requires {async fn body of floria::entities::imperative::imperative::<impl floria::plugins::PluginContext<Arc<Box<dyn floria::AsyncStore + Send + Sync>>>>::set_property()}: Send
  = note: which requires {coroutine witness@floria::entities::imperative::imperative::<impl floria::plugins::PluginContext<Arc<Box<dyn floria::AsyncStore + Send + Sync>>>>::set_property::{closure#0}}: Send
  = note: which requires {async fn body of floria::entities::imperative::edge::<impl floria::Edge>::set_property<Arc<Box<dyn floria::AsyncStore + Send + Sync>>, problemo::Problems>()}: Send
  = note: which requires {coroutine witness@floria::entities::imperative::edge::<impl floria::Edge>::set_property<Arc<Box<dyn floria::AsyncStore + Send + Sync>>, problemo::Problems>::{closure#0}}: Send
  = note: which requires {async fn body of floria::entities::imperative::property::<impl floria::Property>::prepare<Arc<Box<dyn floria::AsyncStore + Send + Sync>>, problemo::Problems>()}: Send
  = note: which requires {coroutine witness@floria::entities::imperative::property::<impl floria::Property>::prepare<Arc<Box<dyn floria::AsyncStore + Send + Sync>>, problemo::Problems>::{closure#0}}: Send
  = note: which requires {async fn body of floria::data::expression::evaluate::<impl floria::Expression>::evaluate<Arc<Box<dyn floria::AsyncStore + Send + Sync>>, FailFast>()}: Send
  = note: which requires {coroutine witness@floria::data::expression::evaluate::<impl floria::Expression>::evaluate<Arc<Box<dyn floria::AsyncStore + Send + Sync>>, FailFast>::{closure#0}}: Send
  = note: which requires Pin<Box<{async fn body of floria::data::expression::evaluate::<impl floria::Expression>::evaluate<Arc<Box<dyn floria::AsyncStore + Send + Sync>>, FailFast>()}>>: Send
  = note: which requires Box<{async fn body of floria::data::expression::evaluate::<impl floria::Expression>::evaluate<Arc<Box<dyn floria::AsyncStore + Send + Sync>>, FailFast>()}>: Send
  = note: which requires std::ptr::Unique<{async fn body of floria::data::expression::evaluate::<impl floria::Expression>::evaluate<Arc<Box<dyn floria::AsyncStore + Send + Sync>>, FailFast>()}>: Send
  = note: which requires {async fn body of floria::data::expression::evaluate::<impl floria::Expression>::evaluate<Arc<Box<dyn floria::AsyncStore + Send + Sync>>, FailFast>()}: Send
  = note: which requires {coroutine witness@floria::data::expression::evaluate::<impl floria::Expression>::evaluate<Arc<Box<dyn floria::AsyncStore + Send + Sync>>, FailFast>::{closure#0}}: Send
  = help: consider adding a manual `impl` of auto traits like `Send` for intermediate types, if auto traits are involved
  = help: or consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`floriad`)
  = note: this lint is attached to the whole crate and can't be disabled on a per-function basis
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #159228 <https://github.com/rust-lang/rust/issues/159228>
  = note: `#[warn(recursion_depth_exceeding_limit)]` (part of `#[warn(future_incompatible)]`) on by default
```

**Comment by sp1ff** [comment]
I'm getting bitten by this as well (could be related to the ScyllaDB issues mentioned above):
```
   |
200 | /         Box::pin(async move {
201 | |             create_scylladb_schema(session, include_str!("../../schemas/0.cql"), 0)
202 | |                 .await
203 | |                 .context(CreateSchemaSnafu)
204 | |         })
    | |__________^
    |
    = note: which requires std::boxed::Box<{async block@indielinks/src/bin/indielinks-schemas.rs:200:18: 200:28}>: std::ops::CoerceUnsized<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>> + std::marker::Send>>
    = note: which requires {async block@indielinks/src/bin/indielinks-schemas.rs:200:18: 200:28}: std::marker::Unsize<dyn futures::Future<Output = std::result::Result<(), Error>> + std::marker::Send>
    = note: which requires {async block@indielinks/src/bin/indielinks-schemas.rs:200:18: 200:28}: std::marker::Send
    = note: which requires {coroutine witness@indielinks/src/bin/indielinks-schemas.rs:200:18: 200:28}: std::marker::Send
    = note: which requires impl futures::Future<Output = std::result::Result<(), indielinks::scylla::Error>>: std::marker::Send
    = note: which requires {async fn body of indielinks::scylla::create_schema()}: std::marker::Send
    = note: which requires {coroutine witness@indielinks::scylla::create_schema::{closure#0}}: std::marker::Send
    = note: which requires impl futures::Future<Output = std::result::Result<(), indielinks::scylla::Error>>: std::marker::Send
    = note: which requires {async fn body of indielinks::scylla::execute_cql<&str>()}: std::marker::Send
    = note: which requires {coroutine witness@indielinks::scylla::execute_cql<&str>::{closure#0}}: std::marker::Send
    = note: which requires futures::stream::Collect<futures::stream::Then<futures::stream::Iter<std::iter::FilterMap<std::str::Split<'_, &str>, {closure@indielinks::scylla::execute_cql<&str>::{closure#0}::{closure#0}}>>, impl futures::Future<Output = std::result::Result<_, _>>, {closure@...}>, _>: std::marker::Send
    = note: which requires... <snip>
```
I'm a bit confused as to how to address this, even after reading "How to best fix this". Am I expected to manually implement `Send` on all of these (some of which I do not control)? How would I implement `Send` on, for instance, ` {async fn body of indielinks::scylla::execute_cql<&str>()}`?

**Comment by ijackson** [comment]
We are having this trouble in Arti, too.  Here's the ticket I just filed there https://gitlab.torproject.org/tpo/core/arti/-/work_items/2715

Again, this seems to be simply a lot of `async` blocks.  There aren't any types we could `unsafe impl Send` for I think.

Maybe this change should be accompanied by an increase in the default recursion limit, to reduce regressions?  Unless there is a better way to fix it.

Also: unfortunately, to get a clean build, we have to suppress this warning.  Will this mean that future upstream crater runs which try to detect this lossage will fail to spot our crates?

**Comment by Kobzol** [comment]
https://github.com/rust-lang/rust/pull/162275 might be able to fix this for async code.

**Comment by maxdexh** [comment]
The recommendation to manually write out auto trait impls might itself lead to breakage. See #162850

**Comment by tliron** [comment]
I'd like to very humbly recommend that the warning be changed until this issue is addressed.

Right now it is this:
```
warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
```
This is scary. It implies that programmers are doing something fundamentally wrong, when in fact it can and does pop up through correct and normal use of Rust, and indeed in some cases (resolving futures created by `async`) there is no workaround. A friendlier warning would be something like "this is a known issue and we are working on it; in the meantime try A, B, or C".

I understand that it's a "new resolver" only warning, but relyng on nightly is unfortunately not optional for many of us.
